File: src/algorithms/treeandgraph/ValidateBinarySearchTree.java
package algorithms.treeandgraph; 

import java.util.*; 

public class ValidateBinarySearchTree {
    /* Function: Check if the tree is a valid BST. 
     * Solution: Cannot use inorder traversal when the defination of BST is left<=root<right. 
     */ 
    
    public boolean isValidBST(TreeNode root) {
        if (root == null) {
            return true; 
        }
        
        ResultType res = checkBST(root); 
        if (res == null) {
            return false; 
        }
        return true; 
    }
    
    private ResultType checkBST(TreeNode root) { // root cannot be null
        ResultType left = root.left == null ? new ResultType(root.val, root.val - 1) : checkBST(root.left); 
        ResultType right = root.right == null ? new ResultType(root.val + 1, root.val) : checkBST(root.right); 
        
        // If as the ITA describes, root.val=left is also valid
        if (left == null || right == null || root.val <= left.max || root.val >= right.min) { 
            return null; 
        }
        
        return new ResultType(left.min, right.max); 
    }
    
    private class ResultType {
        protected int min, max; 
        public ResultType(int min, int max) {
            this.min = min; 
            this.max = max; 
        }
    }
}

File: src/CPP/src/leetcode/ValidateBinarySearchTree.cpp
/*
 * [Source] https://leetcode.com/problems/validate-binary-search-tree/
 * [Difficulty]: Medium
 * [Tag]: Tree
 * [Tag]: Depth-first Search
 */

#include <iostream>

using namespace std;

// [Solution]: Use a structure, return 1. tree valid, 2. smallest value, 3. largest value. Then use recursive call.
// [Corner Case]:
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/treeandgraph/ValidateBinarySearchTree.java
 */

// [Solution]: in-order traverse the tree, and check.
/* Java solution
public class Solution {
    public boolean isValidBST(TreeNode root) {
		if(root==null) return true;
		if(root.left==null && root.right==null) return true;
		int prev=Integer.MIN_VALUE; 
		TreeNode ptr=root;
		TreeNode tmp=null;
		while(ptr!=null){
			if(ptr.left==null){
				if(ptr.val<=prev) return false;
				prev=ptr.val;
			    ptr=ptr.right;
			}else{
				tmp=ptr.left;
				while(tmp.right!=null && tmp.right!=ptr){
					tmp=tmp.right;
				}
				if(tmp.right==ptr){
					if(ptr.val<=prev) return false;
					prev=ptr.val;
					tmp.right=null;
					ptr=ptr.right;
				}else{
					tmp.right=ptr;
					ptr=ptr.left;
				}
			}
		}
		return true;
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
